PerimeterXModule/PxModule.cs

```diff
@@ -232,6 +250,47 @@ private void Application_BeginRequest(object source, EventArgs e)
 			}
 		}
 
+		private bool IsTelemetryCommandValid(HttpContext applicationContext)
+		{
+			//extract header value and decode it from base64 string
+			string headerValue = applicationContext.Request.Headers[PxConstants.ENFORCER_TELEMETRY_HEADER];
+			string decodedString = Encoding.UTF8.GetString(Convert.FromBase64String(headerValue));
+
+			//value is in the form of timestamp:hmac_val
+			string[] splittedValue = decodedString.Split(new Char[] { ':' });
+			if (splittedValue.Length != 2)
+			{
+				PxLoggingUtils.LogDebug("Malformed header value - " + PxConstants.ENFORCER_TELEMETRY_HEADER + " = " + headerValue);
+				return false;
+			}
+
+			//timestamp
+			DateTime expirationTime = (new DateTime(1970, 1, 1)).AddMilliseconds(double.Parse(splittedValue[0]));
+			if ((expirationTime - DateTime.UtcNow).Ticks < 0)
+			{
+				//commmand is expired
+				PxLoggingUtils.LogDebug("Telemetry command is expired");
+				return false;
+			}
+
+			//check hmac integrity
+			string generatedHmac = BitConverter.ToString(new HMACSHA256(cookieKeyBytes).ComputeHash(Encoding.UTF8.GetBytes(splittedValue[0]))).Replace("-","");
+			if(generatedHmac != splittedValue[1].ToUpper())
+			{
+				PxLoggingUtils.LogDebug("hmac validation failed. original = " + splittedValue[1] + ", generated = " + generatedHmac);
+				return false;
+			}
+
+			return true;
+		}
+
+		private bool IsTelemetryCommand(HttpContext applicationContext)
+		{
+			//we got Telemetry command request
```

Comment:
I think this can be a part of the `IsTelemetryCommandValid` method. You can add these two lines as the first lines there.

Reply:
done